### ISSUE TITLE
TransactionPublisher shouldn't stream intermediary results prior to committing the transaction

### ISSUE THREAD

**Issue by zenios** [jOOQ/jOOQ#16401]
### Expected behavior

TransactionPublisher behaves differently with regards to the blocking transaction handling of jooq.

In blocking handling anything that is executed inside the TransactionalCallable is commited and after that anything outside the callable can see the changed data if queried again against the database.

In transaction publisher commit happens onComplete which comes after on next which results in downstream publishers getting results before the transaction is commited

https://github.com/jOOQ/jOOQ/blob/f7fe2b866aabb972df0ba35a1031cc8655577e34/jOOQ/src/main/java/org/jooq/impl/R2DBC.java#L756

### Actual behavior

Transaction is commited prior to onNext thus behaving exactly like the blocking counterpart. 

Regarding this i am not sure if transactionPublisher is the optimal. Maybe it had to be a single ensuring that only one result can be returned

### Steps to reproduce the problem

It is really simple to reproduce but in case its needed i can provide a reproduction. Actually what you need is a transaction that updates a record and on next fetch the record from the database. You will see that the record is still the old because the transaction has not commited yet.

### jOOQ Version

JOOQ Professional Edition 3.19.0

### Database product and version

Postgresql 16

### Java Version

Openjdk 17

**Comment by zenios** [comment]
I have been researching this issue further and i think it has to do with rxjava Maybe that does not trigger onComplete when there is a value. It triggers only onNext

Here are some pieces of code that shows the describe behaviour

**Working example**
```
UUID userId = UUID.fromString("38325510-9035-e481-7677-3708287c4fe3");
for (int i = 0; i < 10; i++) {
    String firstName = Integer.toString(i);
    Single.fromPublisher(dslContext.transactionPublisher(configuration -> Single.fromPublisher(configuration.dsl()
                            .update(USERS)
                            .set(USERS.FIRSTNAME, firstName)
                            .where(USERS.ID.eq(userId))
                            .returning(USERS.FIRSTNAME))
                    .map(r -> r.get(USERS.FIRSTNAME))
                    .toFlowable()))
            .flatMap(u -> Single.fromPublisher(dslContext.select(USERS.FIRSTNAME).from(USERS).where(USERS.ID.eq(userId))).map(Record1::component1))
            .blockingSubscribe(fetchedName -> System.out.println("Values are:" + firstName + " " + fetchedName + " " + fetchedName.equals(firstName)));
}


Values are:0 0 true
Values are:1 1 true
Values are:2 2 true
Values are:3 3 true
Values are:4 4 true
Values are:5 5 true
Values are:6 6 true
Values are:7 7 true
Values are:8 8 true
Values are:9 9 true
```


**Non working example**
```
UUID userId = UUID.fromString("38325510-9035-e481-7677-3708287c4fe3");
for (int i = 0; i < 10; i++) {
    String firstName = Integer.toString(i);
    Maybe.fromPublisher(dslContext.transactionPublisher(configuration -> Single.fromPublisher(configuration.dsl()
                            .update(USERS)
                            .set(USERS.FIRSTNAME, firstName)
                            .where(USERS.ID.eq(userId))
                            .returning(USERS.FIRSTNAME))
                    .map(r -> r.get(USERS.FIRSTNAME))
                    .toFlowable()))
            .flatMapSingle(u -> Single.fromPublisher(dslContext.select(USERS.FIRSTNAME).from(USERS).where(USERS.ID.eq(userId))).map(Record1::component1))
            .blockingSubscribe(fetchedName -> System.out.println("Values are:" + firstName + " " + fetchedName + " " + fetchedName.equals(firstName)));
}

Values are:0 9 false
Values are:1 0 false
Values are:2 1 false
Values are:3 2 false
Values are:4 3 false
Values are:5 4 false
Values are:6 5 false
Values are:7 6 false
Values are:8 7 false
Values are:9 8 false
```

**Comment by lukaseder** [comment]
Thanks a lot for your report

> It is really simple to reproduce but in case its needed i can provide a reproduction

If it would be possible to create a reproducer based on this template, that would be greatly appreciated: https://github.com/jOOQ/jOOQ-mcve

It isn't always necessary. Perhaps it really is simple to reproduce from your description, but often, such descriptions omit a seemingly unimportant detail (e.g. the exact version of RxJava or the jdbc driver), so reproduction will fail nonetheless.

**Comment by marcphilipp** [comment]
We've also come across this issue this week and I created a reproducer: https://github.com/marcphilipp/jooq-transaction-publisher-reproducer

It's not using RxJava but Project Reactor because that was used in our context.

**Comment by lukaseder** [comment]
Thanks a lot for the reproducer, @marcphilipp. I'll look into this, soon

**Comment by lukaseder** [comment]
I can reproduce the issue

**Comment by zenios** [comment]
I have completed an implementation based on the findings in this issue and the related R2DBC
cancellation reports.

- Public fork and branch:
  https://github.com/zenios/jOOQ/tree/r2dbc-cancellation-fix
- Full upstream comparison:
  https://github.com/jOOQ/jOOQ/compare/main...zenios:r2dbc-cancellation-fix
- Related reports:
  [#20112](https://github.com/jOOQ/jOOQ/issues/20112),
  [#20114](https://github.com/jOOQ/jOOQ/issues/20114),
  [#15462](https://github.com/jOOQ/jOOQ/issues/15462), and
  [#18438](https://github.com/jOOQ/jOOQ/issues/18438)

GitHub currently reports that new pull requests are limited to repository collaborators, so I am
posting the complete description here for review. If the approach is acceptable, a collaborator
can open the pull request from the public branch above or enable me to submit it.

### Summary

This change closes several gaps in jOOQ's R2DBC cancellation lifecycle:

- retain cancellation sentinels for connection acquisition, statement execution, transaction
  bodies, and late row publishers, so a late `onSubscribe` cannot revive cancelled work;
- atomically close a connection that is delivered after an ordinary query has already been
  cancelled;
- order accepted connection setup against cancellation, so closing cannot race statement
  creation, binding, or execution subscription setup;
- coordinate transaction cancellation across acquisition, `BEGIN`, body execution, `COMMIT`,
  `ROLLBACK`, and connection close;
- cancel the transaction body and roll back when cancellation wins before `COMMIT`, while allowing
  an already-started `COMMIT` to finish before releasing the connection;
- preserve primary failures and attach rollback/close failures as suppressed exceptions, while
  logging cleanup failures that cannot legally be signalled after downstream cancellation;
- propagate finite, incremental, and unbounded downstream demand to the transaction body through a
  serialized drain instead of requesting `Long.MAX_VALUE` unconditionally;
- document that transaction values are emitted before commit and that only source completion
  establishes post-commit visibility.

### Why the transaction publisher cannot make `Mono` or `Maybe` completion-bound

R2DBC and Reactive Streams expose a general `Publisher`; they do not define `Mono` or `Maybe`
semantics. `transactionPublisher` may stream any number of values, so jOOQ forwards `onNext` while
the transaction body is running and commits only after the body completes.

Reactor's `Mono.from(Publisher)` and RxJava's `Maybe.fromPublisher(Publisher)` both accept the first
value and cancel their source. The downstream success is already visible at that point and cannot
be retracted if the resulting cancellation rolls the transaction back. Changing the existing API
to emit only after commit would require buffering an arbitrary stream until commit, potentially
without a bound, and would break its streaming behaviour.

The Javadoc now recommends completion-aware 0..1 adapters:

- Reactor: `Flux.from(publisher).singleOrEmpty()`
- RxJava: `Flowable.fromPublisher(publisher).singleElement()`
- RxJava exactly-one: `Single.fromPublisher(publisher)`

### Cancellation boundaries

Reactive Streams cancellation is cooperative. jOOQ can request cancellation and ensure that late
subscriptions are cancelled, but it cannot guarantee that connection acquisition, an already
running statement, or a transaction-control command stops immediately. During an active body,
jOOQ cancels the body and requests rollback. An in-progress `BEGIN` is allowed to complete before
rollback, and an in-progress `COMMIT` is allowed to complete before connection release. The R2DBC
driver determines exact wire-protocol draining and command sequencing, so statement and lock
timeouts remain necessary.

### Verification

The downstream patch suite contains 31 focused deterministic tests:

- 11 query cancellation tests covering late connection acquisition, late connection delivery,
  cancellation during statement setup, setup failure, connection-close failure, statement
  execution, result delivery, and row subscription;
- 20 transaction tests covering acquisition/`BEGIN`/body/`COMMIT` cancellation, rollback and close
  failures, delayed body subscription, finite and unbounded demand, reentrant demand,
  finite-to-unbounded promotion, synchronous-stream stack safety, and completion ordering.

The same implementation was also exercised by 17 EmbeddedPostgres persistence integration tests,
including active-statement cancellation, pool-size-one connection reuse, and commit timing.

A warmed zero-I/O benchmark measured the accepted-connection handoff ordering at approximately 14
ns and 32 bytes per synthetic transaction compared with the preceding patch revision. This is a
once-per-query cost, not a per-row cost. The earlier transaction demand coordination measured
approximately 0.8 ns per item on a 100-item synchronous body. Real connection and database latency
dominates both costs.

The public jOOQ checkout does not contain the upstream private regression suite. Equivalent cases,
especially the deterministic #20112 MCVE and the transaction phase races, should be added there by
the maintainers.

### Issue mapping

- Fixes #20112: late statement/result/row subscriptions are cancelled and cannot poison an
  apparently idle pooled connection.
- Fixes #20114: documents eager transaction values, `Mono`/`Maybe` cancellation, and the
  completion-aware alternatives requested by the reporter.
- Addresses #15462: cancellation propagation and late connection release are substantially
  improved, but its original real-driver/pool reproducer should be rerun before closing it.
- Addresses #18438: the same cancellation family is hardened, but the Kotlin Flow/pool reproducer
  should be rerun before closing it.
- Addresses #16401: the eager behaviour remains intentional; the documentation now explains why
  the existing streaming API cannot behave like a completion-bound single-result API.
- Hardens the connection-acquisition cancellation introduced for #17094.

P.S. This PR would not have been possible without the help of AI.

**Comment by lukaseder** [comment]
Thanks a lot for your suggestion. I'll have a look at this soon.

**Comment by lukaseder** [comment]
@zenios: Trying to apply your patch against our integration tests creates tons of regressions, which I will not document here. As this seems to have been generated by AI, mostly, I don't see the value of reviewing the patch as cost/benefit is low compared to simply asking Claude (or similar) to write a patch myself.

**Comment by zenios** [comment]
Feel free but i remember running the integration tests and i dont recall regressions.

**Comment by lukaseder** [comment]
> Feel free but i remember running the integration tests and i dont recall regressions.

You don't have access to our integration tests, so I'm not sure what you mean.

**Comment by lukaseder** [comment]
Looking again into the reproducer by @marcphilipp. This also doesn't produce the issue:

```java
Publisher<Long> transactionPublisher = ctx.transactionPublisher(txn ->
        updateValue(txn.dsl(), id, newValue)
                .then(Mono.delay(Duration.ofMillis(100))));

String downstreamValue = Mono.from(transactionPublisher)
        .then(selectValue(ctx, id))
        .block();

assertEquals("downstream saw state from before transaction", newValue, downstreamValue);
```

Note that I switched from using `concatWith()` to `then()` inside of the transaction.

**Comment by zenios** [comment]
> > Feel free but i remember running the integration tests and i dont recall regressions.
> 
> You don't have access to our integration tests, so I'm not sure what you mean.

I think this is a timing artifact.

concatWith emits the update result, then waits 100 ms before completing. That gives the downstream SELECT time to read before commit.

then discards the update result and emits after the delay, immediately before completion. Commit can now finish before the SELECT reads the row.

So the change removes the delay that exposed the issue, but doesn’t guarantee the read happens after commit.

That’s why my [documentation change](https://github.com/zenios/jOOQ/commit/bcb9dcc63156c8ac20c6654ae6fef73b45e6fc51) clarified that callers must wait for the transaction publisher to complete successfully, rather than just receive its first value.

**Comment by lukaseder** [comment]
~I don't think that explains it. The following code clearly sequences the publishers in a way where the transaction needs to complete *before* selecting the value, so this is not a client issue, at least not in this reproducer.~

```java
Mono.from(transactionPublisher)
        .then(selectValue(ctx, id))
```

Edit: Nevermind, you're right. The `UPDATE` emits the first (and only) value requested by the caller prior to calling `COMMIT`, which still happens on the thread, but too late. So, the solution could be to buffer all results, and forward them only once `COMMIT` has happened.

**Comment by lukaseder** [comment]
See also parallel discussion here:

- https://github.com/jOOQ/jOOQ/issues/20114#issuecomment-5600879109

I wonder if the streaming behaviour is the actual bug here, which can be fixed easily by buffering the transaction results.

**Comment by lukaseder** [comment]
What makes this even more weird is this:

```java
System.out.println(Flux.from(DSL.using(cf, dialect)
   .transactionPublisher(ctx -> {
       Flux<Integer> f1 = Flux.from(ctx.dsl().insertInto(T_AUTHOR).columns(T_AUTHOR.ID, T_AUTHOR.LAST_NAME).values(3, "X"));
       Flux<Integer> f2 = f1.thenMany(ctx.dsl().select(T_AUTHOR.ID).from(T_AUTHOR).orderBy(T_AUTHOR.ID.desc())).map(Record1::value1);
       Flux<Integer> f3 = f2.concatWith(Mono.delay(Duration.ofMillis(100)).then(Mono.empty()));
       Flux<Integer> f4 = f3.concatWith(ctx.dsl().insertInto(T_AUTHOR).columns(T_AUTHOR.ID, T_AUTHOR.LAST_NAME).values(4, "X"));

       return f4;
   }))
    .blockFirst()
);

System.out.println(
    Flux.from(DSL.using(cf, dialect).select(T_AUTHOR.ID).from(T_AUTHOR)).map(Record1::value1).collectList().block()
);

Thread.sleep(1000);

System.out.println(
    Flux.from(DSL.using(cf, dialect).select(T_AUTHOR.ID).from(T_AUTHOR)).map(Record1::value1).collectList().block()
);
```

This prints the following  `DEBUG` log output:

```
14:07:20,504  INFO [DatabaseSetup                 ] [72] - Open R2DBC connections (open) : 1
14:07:20,506 DEBUG [LoggingConnection             ] [72] - Connection::beginTransaction: org.jooq.impl.R2DBC$TransactionSubscription$1$1@34798e21
14:07:20,965 DEBUG [LoggingConnection             ] [72] - Connection::createStatement: insert into "public"."t_author" ("id", "last_name") values ($1, $2)
14:07:20,973 DEBUG [LoggingStatement              ] [72] - Statement::execute       
14:07:21,004 DEBUG [LoggingConnection             ] [72] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author" order by "public"."t_author"."id" desc
14:07:21,007 DEBUG [LoggingStatement              ] [72] - Statement::execute       
3
14:07:21,038  INFO [DatabaseSetup                 ] [74] - Open R2DBC connections (open) : 2
14:07:21,039 DEBUG [LoggingConnection             ] [74] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author"
14:07:21,039 DEBUG [LoggingStatement              ] [74] - Statement::execute       
14:07:21,042 DEBUG [LoggingConnection             ] [74] - Connection::close        
14:07:21,046  INFO [DatabaseSetup                 ] [74] - Open R2DBC connections (close): 1
[1, 2]
14:07:21,122 DEBUG [LoggingConnection             ] [104] - Connection::createStatement: insert into "public"."t_author" ("id", "last_name") values ($1, $2)
14:07:21,122 DEBUG [LoggingStatement              ] [104] - Statement::execute       
14:07:21,133 DEBUG [LoggingConnection             ] [72] - Connection::rollbackTransaction
14:07:21,135 DEBUG [LoggingConnection             ] [72] - Connection::close        
14:07:21,135  INFO [DatabaseSetup                 ] [72] - Open R2DBC connections (close): 0
14:07:22,061  INFO [DatabaseSetup                 ] [76] - Open R2DBC connections (open) : 1
14:07:22,061 DEBUG [LoggingConnection             ] [76] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author"
14:07:22,061 DEBUG [LoggingStatement              ] [76] - Statement::execute       
14:07:22,062 DEBUG [LoggingConnection             ] [76] - Connection::close        
14:07:22,063  INFO [DatabaseSetup                 ] [76] - Open R2DBC connections (close): 0
[1, 2]
```

Observe how the transaction leaks its internal value (`3`, the inserted ID) but is rolled back, because of the cancellation. This seems kind of reasonable. When we call `Flux::blockLast`, then the whole `transactionPublisher` content is emitted and the transaction is committed. This is inconsistent with using `Mono`:

```java
System.out.println(Mono.from(DSL.using(cf, dialect)
   .transactionPublisher(ctx -> {
       Flux<Integer> f1 = Flux.from(ctx.dsl().insertInto(T_AUTHOR).columns(T_AUTHOR.ID, T_AUTHOR.LAST_NAME).values(3, "X"));
       Flux<Integer> f2 = f1.thenMany(ctx.dsl().select(T_AUTHOR.ID).from(T_AUTHOR).orderBy(T_AUTHOR.ID.desc())).map(Record1::value1);
       Flux<Integer> f3 = f2.concatWith(Mono.delay(Duration.ofMillis(100)).then(Mono.empty()));
       Flux<Integer> f4 = f3.concatWith(ctx.dsl().insertInto(T_AUTHOR).columns(T_AUTHOR.ID, T_AUTHOR.LAST_NAME).values(4, "Y"));

       return f4;
   }))
    .block()
);

System.out.println(
    Flux.from(DSL.using(cf, dialect).select(T_AUTHOR.ID).from(T_AUTHOR)).map(Record1::value1).collectList().block()
);

Thread.sleep(1000);

System.out.println(
    Flux.from(DSL.using(cf, dialect).select(T_AUTHOR.ID).from(T_AUTHOR)).map(Record1::value1).collectList().block()
);
```

Producing:

```
14:11:43,281  INFO [DatabaseSetup                 ] [74] - Open R2DBC connections (open) : 1
14:11:43,283 DEBUG [LoggingConnection             ] [74] - Connection::beginTransaction: org.jooq.impl.R2DBC$TransactionSubscription$1$1@3abf65e6
14:11:43,436 DEBUG [LoggingConnection             ] [74] - Connection::createStatement: insert into "public"."t_author" ("id", "last_name") values ($1, $2)
14:11:43,437 DEBUG [LoggingStatement              ] [74] - Statement::execute       
14:11:43,448 DEBUG [LoggingConnection             ] [74] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author" order by "public"."t_author"."id" desc
14:11:43,449 DEBUG [LoggingStatement              ] [74] - Statement::execute       
3
14:11:43,472  INFO [DatabaseSetup                 ] [76] - Open R2DBC connections (open) : 2
14:11:43,472 DEBUG [LoggingConnection             ] [76] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author"
14:11:43,472 DEBUG [LoggingStatement              ] [76] - Statement::execute       
14:11:43,474 DEBUG [LoggingConnection             ] [76] - Connection::close        
14:11:43,474  INFO [DatabaseSetup                 ] [76] - Open R2DBC connections (close): 1
[1, 2]
14:11:43,563 DEBUG [LoggingConnection             ] [104] - Connection::createStatement: insert into "public"."t_author" ("id", "last_name") values ($1, $2)
14:11:43,563 DEBUG [LoggingStatement              ] [104] - Statement::execute       
14:11:43,569 DEBUG [LoggingConnection             ] [74] - Connection::commitTransaction
14:11:43,572 DEBUG [LoggingConnection             ] [74] - Connection::close        
14:11:43,572  INFO [DatabaseSetup                 ] [74] - Open R2DBC connections (close): 0
14:11:44,499  INFO [DatabaseSetup                 ] [78] - Open R2DBC connections (open) : 1
14:11:44,499 DEBUG [LoggingConnection             ] [78] - Connection::createStatement: select "public"."t_author"."id" from "public"."t_author"
14:11:44,500 DEBUG [LoggingStatement              ] [78] - Statement::execute       
14:11:44,501 DEBUG [LoggingConnection             ] [78] - Connection::close        
14:11:44,501  INFO [DatabaseSetup                 ] [78] - Open R2DBC connections (close): 0
[1, 3, 4, 2]
```

Now, there's a background transaction still running, which eventually commits, so no cancellation is being propagated.

We can argue two things:

- In general, a transaction could be seen as a unit of work. It should not emit intermediary state downstream prior to the transaction completing, i.e. until it is known what the outcome is. In this case, both the `Mono` and `Flux` cases are wrong.
- In some cases, it may be desirable for intermediary transaction state to become available outside of the transaction scope, kinda similar to `READ UNCOMMITTED` transaction isolation. I.e. a second transaction (or stream client in this case) can read the transaction's contents even before the commit finalises them.

I feel that the default should be consistent with most people's expectations, i.e. a "unit of work" is atomic and intermediary results aren't being streamed. This is consistent with people's bug reports here. A new `Settings` will allow for reverting to the existing behaviour, where `Mono` still needs to be fixed, though. If a `transactionPublisher` in a "streaming mode" is consumed only half way through and then cancelled, it should roll back

**Comment by lukaseder** [comment]
A follow-up task allows for restoring behaviour similar to the status quo, in case anyone really prefers the streaming style:

- https://github.com/jOOQ/jOOQ/issues/20199

**Comment by lukaseder** [comment]
The status quo is:

- Inconsistent with JDBC and coroutines
- Undefined
- Buggy (publisher cancellation may lead to rollback or background transaction, depending on usage)

Nevertheless, users may rely on the status quo. Especially in the case where cancellation already rolls back a transaction, the change would be breaking. In the case where the commit happens in the background, the change could be seen as breaking subtly, as it no longer happens in the background, but instead, stream operations are now sequenced differently, more predictably.

As such, I will not backport this fix, but instead document the behaviour as undefined in 3.19, 3.20, and 3.21.

**Comment by lukaseder** [comment]
Fixed in versions:

- 3.22.0

Documentation backport to versions:

- 3.21.9 (https://github.com/jOOQ/jOOQ/issues/20200)
- 3.20.20 (https://github.com/jOOQ/jOOQ/issues/20201)
- 3.19.39 (https://github.com/jOOQ/jOOQ/issues/20202)

**Comment by lukaseder** [comment]
Reopening as the current implementation might cause a Rule 1.1 violation ("MUST NOT emit more onNext signals than requested")
